### DISCUSSION TITLE
Policy to load stubs for PHP extensions (core, bundled, external, PECL)

### DISCUSSION THREAD

**Discussion by alies-dev** [vimeo/psalm#8879] [Psalm development (AKA internals)]
What is the policy for PHP extensions?

When I run Psalm I see something like this:
```
Target PHP version: 8.1 (inferred from composer.json) Extensions enabled: dom, pdo (unsupported extensions: bcmath, curl, filter, gd, hash, iconv, intl, json, mbstring, openssl, pcntl, pdo_mysql, redis, tokenizer, zip)
```

composer.json file example:
```json
{
    "require": {
        "php": "^8.1",
        "ext-PDO": "*",
        "ext-bcmath": "*",
        "ext-curl": "*",
        "ext-dom": "*",
        "ext-filter": "*",
        "ext-gd": "*",
        "ext-hash": "*",
        "ext-iconv": "*",
        "ext-intl": "*",
        "ext-json": "*",
        "ext-mbstring": "*",
        "ext-openssl": "*",
        "ext-pcntl": "*",
        "ext-pdo_mysql": "*",
        "ext-redis": "*",
        "ext-tokenizer": "*",
        "ext-zip": "*",
...
}
```

## ext types

There are [few types of ext](https://www.php.net/manual/en/extensions.membership.php):
1. **core** (These are not actual extensions. They are part of the PHP core and cannot be left out of a PHP binary with compilation options.)
2. **bundled** (These extensions are bundled with PHP.)
3. **external** (These extensions are bundled with PHP but in order to compile them, external libraries will be needed.)
4. **PECL** (These extensions are available from [PECL](https://pecl.php.net/). They may require external libraries.)

Additionally, different PHP versions have different bundled extensions, e.g. [sodium](https://www.php.net/manual/en/sodium.installation.php) was PECL ext and became bundled as of PHP 7.2.0.

## how Psalm loads stubs for extensions
 - extract from composer.json
 - using config: `enableExtensions` key

Then load stubs:
https://github.com/vimeo/psalm/blob/6b3dd568504a1e13f3ffdd21d8a2b80bc9a060b3/src/Psalm/Config.php#L2138-L2143

Recently a new way has been introduced ([probably](https://github.com/vimeo/psalm/issues/7912) to load ext stubs on https://psalm.dev/ snippets): using `extension_loaded()` function, first time used in https://github.com/vimeo/psalm/pull/7614 . I think it's save to treat ext-redis as a regular extension.

## proposals

1. Do not confuse users by a list of "unsupported extensions" that actually supported (usually core and some bundled one)
2. Reduce options to load extensions: do not use `extension_loaded`, instead, use existing 2 ways of doing this (using composer.json is consider as the best practice for doing this). Load all available ext stubs on psalm.dev (unless there are some conflicting). (a PR as an example https://github.com/vimeo/psalm/pull/8881)

**Comment by weirdan** [comment]
> do not use `extension_loaded()`

Sounds good to me

**Comment by orklah** [comment]
I'm not too convinced myself.

1. I agree that having unsupported extensions is confusing, but this is due to the fact that Psalm's stubs are not yet all complete. We could just drop that phrase to avoid this confusion
2. dropping extension_loaded seems a clear downside for casual users (forcing them to manually set the configuration) with a vague upside (having less entry points)

Dropping extension_loaded now,after we just released Psalm 5 in which the config can be set for the first time, will guarantee that CIs will fail everywhere because of missing extensions for no good reason.
If we want to do it anyway, we'll have to soften the blow:
- first with a warning that any loaded extension not configured in either composer or config will be dropped in the future
- second with an automatic configuration on creation of the config file (through --init) that creates the configuration with any loaded extension at this moment

**Reply by alies-dev** [reply]
Agree, we should not break existing integrations. I've created a new PR taking into account your suggestions: https://github.com/vimeo/psalm/pull/8885
